src/Storage/Notmuch.hs

```diff
@@ -218,6 +218,23 @@ threadToThread m = do
       <*> Notmuch.threadTotalMessages m
       <*> Notmuch.threadId m
 
+reloadTags
+    :: (MonadError Error m, MonadIO m, HasId a, ManageTags a)
+    => FilePath -> a -> m a
+reloadTags fp item = withDatabaseReadOnly fp go
+  where
+    go db = getThread db (getId item) >>= Notmuch.tags >>= pure . flip setTags item
+
+class HasId a where
```

Comment:
I think this class is a bad idea, because `ThreadId` and `MessageId` - although
they are type synonyms now for `ByteString`, could eventually become different
types via `newtype`.

If you want to keep the class so there is a common access method `getId`, you can
use an associated type:

```haskell
class HasId a where
  type IdType a
  getId :: a -> IdType a

instance HasId NotmuchMail where
  type IdType NotmuchMail = MessageId
  getId = view mailId

...
```

This requires the `TypeFamilies` language extension.

I'm inclined to suggest `getId :: Getter a (IdType a)` too so it can be composed
with other optics.  But I am not fussy about that.

Reply:
Hm... `getId` now returns a type of `IdType a`, but `getThread` wants a `ByteString`. I'm now a bit puzzled how I get my Bytestring out of `IdType` again. Would I need to define a data type or type alias to define another function to pull out the Bytestring?

Reply:
Actually, I'm realising that this can not be generalised. The function calls `getThread` which in turn queries the database with the given thread `id`. While technically it won't fail, passing in a mail ID would return no results and hence doesn't really make sense. Perhaps `getThread` would have to be abstracted so that you can pass in a message- or thread id.

Reply:
I removed the class now, since anything which could be abstracted away is perhaps something for another patch.